### PR TITLE
Add GitHub Actions workflows targeting pip-installed dependencies

### PR DESCRIPTION
**Pull request by kitchoi**
This PR adds GitHub Actions workflows for testing under different configurations using pip only.

This PR excludes testing against PyQt5 and PyQt + Python 3.6 as they will be tested with EDM packages. A preview of those jobs can be seen here: https://github.com/enthought/traitsui/actions/runs/385734961
Those jobs require a GitHub Action for setting up EDM, which will be provided from a separate GitHub repository.

Jobs included:
- Styling checks against Python 3.6 and Python 3.9
- Tests with minimal dependencies
- Tests against PySide2 on Ubuntu / Windows / OSX
- Tests against wxPython 4 on Windows
- Tests against Traits 6
- Tests against master of Traits and Pyface
- Tests for demo examples with Qt and Ubuntu

Currently I set all the jobs to be run on push, so that I can test this PR. However, prior to merging, we should decide which of these are run per-PR, and which are to be run as scheduled cron jobs.

Proposal:

Per-PR
- Jobs for PyQt5 and PyQt with EDM on three platforms (not in this PR)     ---- 6 jobs
- Styling checks against Python 3.6 and Python 3.9.                                      ---- 2 jobs
- Tests with minimal dependencies                                                                   ---- 1 job
- Tests against wxPython 4 on Windows                                                          ---- 1 job
- Tests against Traits 6                                                                                       ---- 1 job
- Tests for demo examples with Qt and Ubuntu                                               ---- 1 job

Total: 12 jobs

Cron:
- Tests against master of Traits and Pyface on three platforms                     ---- 3 job
- Tests against PySide2 from PyPI on three platforms                                    ---- 3 jobs

Total: 6 jobs

I will reorganize the jobs according to what we decide to run as scheduled cron jobs.

Note:
For wxPython jobs, the test suite for Ubuntu has been allowed to fail on Travis, and it fails similarly here too. I managed to run the tests and reproduce the errors with GitHub Actions, using an Action that install a prebuilt wheel from wxPython.org with dynamically computed URL. However, given the job is failing, and GitHub Actions has yet to support something like `allow-failure` on Travis (see https://github.com/actions/toolkit/issues/399), I removed the job eventually. See this commit for resurrecting the code: a3bc4f4c93e6418c816140368ee8f41b300c7033

Related: https://github.com/enthought/traitsui/issues/1387

### CODE REVIEW

**Review by mdickinson — APPROVED** [review]
LVGTM. Let's merge and tweak as the need arises.

**Comment by mdickinson** [comment]
> we should decide which of these are run per-PR, and which are to be run as scheduled cron jobs

I'd go with a bare minimum for per-PR runs: wxPython 4 and PySide2/Qt5 toolkits on Python 3.6, across 3 platforms (for a total of 6 jobs). Or do you think we should include the null toolkit, too?

**Comment by mdickinson** [comment]
> I'd go with a bare minimum for per-PR runs:

Ah, sorry; we'd definitely want the style check, too.

File: .github/workflows/test-with-pip.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+          run: python -m unittest discover -v traitsui
+          working-directory: ${{ runner.temp }}
+
+  # Tests against wxPython from PyPI
+  # On OSX:
+  #   - wxPython complains about 'This program needs access to the screen'.
+  # On Ubuntu:
+  #   - Prebuilt wheels from wxpython.org are not compatible for our Python 3.8
+  #     and 3.9 setup. We can build from source and do something about
+  #     it taking a long time to build.
```
Comment (kitchoi) [inline .github/workflows/test-with-pip.yml:46]: The first sentence was actually only true when I tested with wxPython 4.1.0. wxPython 4.1.1 was released after this comment was written and the wheels for wxPython 4.1.1 are compatible with Python 3.8 and Python 3.9 here (I tested them). I will update this comment.

**Review by kitchoi — COMMENTED** [review]
> I'd go with a bare minimum for per-PR runs: wxPython 4 and PySide2/Qt5 toolkits on Python 3.6, across 3 platforms (for a total of 6 jobs). Or do you think we should include the null toolkit, too?
> Ah, sorry; we'd definitely want the style check, too.

Sounds good to me!  I will update the `on` triggers accordingly.

The 'Minimal setup' (test with minimal dependencies) is testing against the null toolkit, so let's keep that for per-PR runs. It is true that the 'Minimal setup' will skip tests for ArrayEditor and DataFrameEditor by virtue of not having NumPy and Pandas, rather than by virtue of having no actual GUI toolkits. However, it seems a bit lavish to expand the matrix here for the null toolkit.

I forgot about the the CI jobs for ets-demo 😓 .  I will do that in a separate PR. Just wanted to mention that this will add a few per-PR jobs, but we can add filters on the paths such that the jobs are skipped if nothing in ets-demo has changed.

**Comment by kitchoi** [comment]
Thanks @mdickinson!

Merging. We will tweak this as we go. I opened #1416 to track ongoing migration.

**Pull request merged**